pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -775,13 +775,26 @@ func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(ctx context.Con
 			containerResults <- killContainerResult
 		}(container)
 	}
-	wg.Wait()
-	close(containerResults)
+	killContainerCompletedSignal := make(chan struct{}, 1)
```

Comment:
I don't understand why this is necessary - killContainer is taking a context and cancellation there should be sufficient.

Is there evidence it doesn't work?

Reply:
The first paragraph confuses me.

When ctx is cancelled, killContainer should exit rapidly.  Is this not the case?  If it isn't, we should be fixing the bugs at that level (not in this method)

Reply:
> The first paragraph confuses me.
> 
> When ctx is cancelled, killContainer should exit rapidly. Is this not the case? If it isn't, we should be fixing the bugs at that level (not in this method)

As I understand it, although the killContainer function passes ctx, but without <-ctx.Done() to receive the exit value, the CRI in blocking will not end, right?
If even killContainer receives the exit signal from ctx, it should return some result information. So my change is to better be able to return the latest information back

Reply:
> The first paragraph confuses me.
> 
> When ctx is cancelled, killContainer should exit rapidly. Is this not the case? If it isn't, we should be fixing the bugs at that level (not in this method)

Maybe I misled you with my previous statement. The correct code logic should be
The ctx passed in killContainer is not able to receive the ctx.cancel signal. It must be received by <-ctx.Done() to do so. Otherwise it still can't exit goroutines
"The signal of ctx.cancel needs to be received by ctx.Done. and not goroutines"

Reply:
> The first paragraph confuses me.
killContainerCompletedSignal The role of this structure is probably not clear to me either.
Because according to the original code, m.killContainer(ctx, ...) And this is a goroutines inside a for. you have to make sure that all containers are executed.
killContainerCompletedSignal is to ensure that all for containers are executed successfully when ctx.Done is not received. If this is not added, the original code wg.wait() will keep blocking and we will not be able to receive ctx.Done successfully

Reply:
>  Is this not the case? If it isn't, we should be fixing the bugs at that level (not in this method)

In the current case, killContainer does not exit quickly when ctx is cancelled.
If you think you shouldn't do ctx.Done in the current method, you should receive ctx.Done in the killContainer function, but this is a grpc call to CRI, and I don't know what this function is in

Reply:
> In the current case, killContainer does not exit quickly when ctx is cancelled.

In that case, we need to change the `killContainer` method so that it does exit quickly when ctx is cancelled.  The expectation is when ctx is cancelled that all methods return quickly - whether that's the CRI (which takes ctx as an argument and should stop the current request) or the methods themselves if they do particularly long loops.

Reply:
> killContainer

So what should we do now, move this ctx.Done reception to the killContainer method or move it to the CRI side?
You can look at line 745 of this file, which is where the killContainer method eventually calls the StopContainer method (which eventually blocks here). So do we copy the current code after 745? Or do we solve it at the CRI?

Reply:
> goroutines

But I think it might be more appropriate to put it in the killContainersWithSyncResult now (since the goroutines are already executed) If we put it in the killContainer function, then we need to modify the StopContainer function to make it goroutines. No matter where the change is made, there must be goroutines to execute and to receive ctx.Done() to do so.
If we put it in CRI, we need goroutines to execute and receive the ctx.Done signal

Reply:
> In that case, we need to change the `killContainer` method so that it does exit quickly when ctx is cancelled. The expectation is when ctx is cancelled that all methods return quickly

@smarterclayton Hi, I recently looked at this part of the code in detail again.
I misunderstood what you meant before (It had to be changed at CRI level). It turns out that you meant "change it in the killContainer function".
This is the latest result of the change I made in the killContainer function and tested it.